parla/task_runtime.py

```diff
@@ -1923,46 +1981,144 @@ def _schedule_tasks(self):
     def _launch_task(self, queue, dev: Device):
         try:
             task = queue.pop()
+
             worker = self._free_worker_threads.pop()  # grab a worker
+
             logger.info(f"[Scheduler] Launching %s task, %r on %r",
                         dev.architecture.id, task, worker)
             worker.assign_task(task)
             logger.debug(f"[Scheduler] Launched %r", task)
+
             for dev in task.req.environment.placement:
-                self.update_launched_task_count(task, dev, 1)
+                self.update_launched_task_count_mutex(task, dev, 1)
         except IndexError:
             # If failed to find available thread, reappend it.
-            queue.append(task)
+            self.enqueue_dev_queue(dev, task)
 
     def _launch_tasks(self):
         """ Iterate through free devices and launch tasks on them
         """
-        #logger.debug("[Scheduler] Launch Phase")
-        with self._monitor:
+        # logger.debug("[Scheduler] Launch Phase")
+        for dev in self._available_resources.get_resources():
+            with self._dev_queue_monitor[dev]:
+                with self._thread_queue_monitor:
+                    if len(self._free_worker_threads) == 0:
+                        break
+                    compute_queue = self._compute_task_dev_queues[dev]
+                    datamove_queue = self._datamove_task_dev_queues[dev]
+                    if len(compute_queue) > 0:
+                        if dev.architecture.id == "cpu" or self.get_launched_compute_task_count(dev) < (self._num_colocatable_tasks + 1):
+                            self._launch_task(compute_queue, dev)
+                    if len(datamove_queue) > 0:
+                        if dev.architecture.id == "cpu" or self.get_launched_datamove_task_count(dev) < (self._num_colocatable_tasks + 1):
+                            self._launch_task(datamove_queue, dev)
+
+    def map_tasks_callback(self):
+        """ This is a callback function for the mapper. Called by WorkerThread and Spawn Decorator to trigger scheduler execution"""
+
+        # Check runtime condition
+        # Are there tasks that have not yet been mapped?
+        # Is the ready queue under a certain threshold?
+        ready_queue_threshold = 50
```

Comment:
not gonna use this?

Reply:
Added it as a potential condition to make the callback run less often. But disabled all of the other checks for debugging. In this commit, all callbacks run as often as they can.